tests/Unit/Parallel/Test_MemoryMonitor.cpp

```diff
@@ -139,8 +173,215 @@ void test_initialize() {
+  ActionTesting::next_action<obs_writer_comp>(runner, 0);
+
+  // MemoryMonitor
+  ActionTesting::emplace_singleton_component_and_initialize<mem_mon_comp>(
+      runner, ActionTesting::NodeId{2}, ActionTesting::LocalCoreId{2}, {});
+
+  runner->set_phase(metavars::Phase::Monitor);
+}
+
+template <typename Component>
+void check_output(const std::string& filename, const double time,
+                  const size_t num_nodes, const std::vector<double>& sizes) {
+  h5::H5File<h5::AccessType::ReadOnly> read_file{filename + ".h5"};
+  INFO("Checking output of " + pretty_type::name<Component>());
+  const auto& dataset =
+      read_file.get<h5::Dat>(mem_monitor::subfile_name<Component>());
+  const std::vector<std::string>& legend = dataset.get_legend();
+
+  size_t num_columns;
+  if constexpr (Parallel::is_group_v<Component>) {
+    // time, size on node 0, size on node 1, ...,  proc of max size, max size,
+    // avg per node
+    num_columns = num_nodes + 4;
+  } else {
+    // time, size on node 0, size on node 1, ..., avg per node
+    num_columns = num_nodes + 2;
+  }
+
+  CHECK(legend.size() == num_columns);
+
+  const Matrix data = dataset.get_data();
+  // We only wrote one line of data for all components
+  CHECK(data.rows() == 1);
+  CHECK(data.columns() == num_columns);
+
+  const double average =
+      alg::accumulate(sizes, 0.0) / static_cast<double>(num_nodes);
+
+  CHECK(data(0, 0) == time);
+  for (size_t i = 0; i < num_nodes; i++) {
+    CHECK(data(0, i + 1) == sizes[i]);
+  }
+  CHECK(data(0, num_columns - 1) == average);
+}  // namespace
+
+template <typename Component>
+void run_group_actions(ActionTesting::MockRuntimeSystem<metavars>& runner,
+                       const std::string& filename, const int num_nodes,
+                       const int num_procs, const double time,
+                       const std::vector<double>& sizes) {
+  INFO("Checking actions of " + pretty_type::name<Component>());
+  size_t num_branches;
+  if constexpr (Parallel::is_group_v<Component>) {
+    num_branches = static_cast<size_t>(num_procs);
+  } else {
+    num_branches = static_cast<size_t>(num_nodes);
+    // Avoid unused parameter warning
+    (void)num_procs;
+  }
+
+  CHECK(ActionTesting::number_of_queued_simple_actions<mem_mon_comp>(
+            runner, 0) == num_branches);
+
+  const auto& mem_holder_tag =
+      ActionTesting::get_databox_tag<mem_mon_comp,
+                                     mem_monitor::Tags::MemoryHolder>(runner,
+                                                                      0);
+
+  // Before we invoke the actions, the map for this component shouldn't exist
+  // because this is the first time we are calling it.
+  const std::string name = pretty_type::name<Component>();
+  CHECK(mem_holder_tag.count(name) == 0);
+
+  for (size_t i = 0; i < num_branches; i++) {
+    // unordered_map<name, unordered_map<time, unordered_map<node/proc, size>>>
+    // Skip 0, otherwise an access error will occur
```

Comment:
I'm a bit confused, would that mean you want the check _after_ the action?

Reply:
So before the first action is invoked on the memory monitor, the map for that time should be empty. And after the last action, the map should also be empty (it gets cleaned up). So if I move the check after the action, I'd have to do something like

```
if (i != num_branches-1) {
  CHECK(...);
}
```

to avoid an access error after the last action. So, it's just a matter of where you want to put the skip.

Reply:
Ah okay, could you maybe rephrase the comment to include that detail? That is, `The map is empty both before and after the last call, so we need to put the CHECK inside an if clause regardless`?